src/Module/Debug/ActivityPubConversion.php

```diff
@@ -134,4 +126,9 @@ function visible_whitespace($s)
 
 		return $o;
 	}
+
+	private function visible_whitespace(string $s): string
```

Comment:
This used to be a quick helper function used only in Debug classes. If you must make it into a full-fledged class method, please move it to the `Content\Text\HTML` class and rename it `visibleWhitespace`.

Reply:
The problem here was the definition of a named function inside a method. This way you can't call the method twice, see https://3v4l.org/Y7iu7 I can also create an anonymous function inside the method to fix this.

If this quick helper function is only for debugging I would keep it as a private method. There is no problem with using a `full-fledged` method for this. I would not recommend to create a new dependency to the `Content\Text\HTML`, because this will extend our API surface that we have to maintain.

Reply:
Interesting, I didn't consider that. However the `rawContent()`, `content()` and `post()` methods are meant to be only called once per request so the issue never arose. Was it stunting the tests by any chance?

As for the dependency to `Content\Text\HTML`, it's the one area where I know how to write extensive unit tests since methods are pure and don't depend on any other service. This makes me confident about using it as a dependency itself.

If it still had been used only in one place, I would have preferred going the private function route, but since it's now used in two different places, I'd rather it being centralized in a single place which I think should be `Content\Text\HTML`.

Reply:
> Was it stunting the tests by any chance?

I'm not sure that this method is called in tests but if it was called twice you definitely would see it.

> As for the dependency to `Content\Text\HTML`, it's the one area where I know how to write extensive unit tests since methods are pure and don't depend on any other service. This makes me confident about using it as a dependency itself.
>
> If it still had been used only in one place, I would have preferred going the private function route, but since it's now used in two different places, I'd rather it being centralized in a single place which I think should be `Content\Text\HTML`.

It is not about the `Content\Text\HTML` class but about the `ActivityPubConversion` class that worries me more. In my opinion this small html code would be best placed in the `view/templates/debug/activitypubconversion.tpl` template file. This should be done in a separate PR.

I will change this code to an anonymous function, if you agree.